fqdn: Add Protocol to DNS Proxy Cache

nathanjsweet opened the pull request:
DNS Proxy indexes domain selectors by port
only. In cases where protocols collide on port
the DNS proxy may have a more restrictive selector than it should because it does not merge port
protocols for L7 policies (only ports).

Refactor all users of the DNS Proxy are updated
to add protocol to any DNS Proxy entries, and all
tests are updated to test for port-protocol
merge errors.

```release-note
fqdn: Fixed bug that caused DNS Proxy to be overly restrictive on allowed DNS selectors.
```

nathanjsweet committed:
fqdn: Add Protocol to DNS Proxy Cache

DNS Proxy indexes domain selectors by port
only. In cases where protocols collide on port
the DNS proxy may have a more restrictive selector
than it should because it does not merge port
protocols for L7 policies (only ports).

All callers of the DNS Proxy are updated
to add protocol to any DNS Proxy entries, and all
tests are updated to test for port-protocol
merge errors.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

pippolo84 commented on pkg/endpoint/policy.go:
```diff
@@ -91,16 +91,16 @@ func (e *Endpoint) getNamedPortEgress(npMap types.NamedPortMultiMap, name string
 // proxyID returns a unique string to identify a proxy mapping,
 // and the resolved destination port number, if any.
```
Nit: docstring should be updated to mention the returned protocol

pippolo84 commented on pkg/fqdn/dnsproxy/helpers.go:
```diff
@@ -13,13 +13,13 @@ import (
 // lookupTargetDNSServer finds the intended DNS target server for a specific
 // request (passed in via ServeDNS). The IP:port combination is
```
Same here regarding the comment

pippolo84 commented on pkg/fqdn/dnsproxy/helpers.go:
```diff
@@ -13,13 +13,13 @@ import (
 	switch addr := (w.LocalAddr()).(type) {
 	case *net.UDPAddr:
-		return addr.IP, uint16(addr.Port), addr.String(), nil
+		return addr.IP, uint16(addr.Port), 17, addr.String(), nil
```
`17` seems kind of a magic number, I think it'd be better to use `uint16(u8proto.UDP)` instead.

pippolo84 commented on pkg/fqdn/dnsproxy/helpers.go:
```diff
@@ -13,13 +13,13 @@ import (
+		return addr.IP, uint16(addr.Port), 17, addr.String(), nil
 	case *net.TCPAddr:
-		return addr.IP, uint16(addr.Port), addr.String(), nil
+		return addr.IP, uint16(addr.Port), 6, addr.String(), nil
```
Same here: `6` -> `uint16(u8proto.TCP)`

pippolo84 commented on pkg/fqdn/dnsproxy/helpers_test.go:
```diff
@@ -19,6 +19,11 @@ import (
+	tcpProto = 6
+	udpProto = 17
+)
+
```
Any particular reason to redefine these here instead of using the constants from `u8proto`? If we want to avoid repeating the type conversion at each call we might as well do:

```
const (
	tcpProto = uint16(u8proto.TCP)
	udpProto = uint16(u8proto.UDP)
)
```

pippolo84 commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -166,17 +166,17 @@ type regexCacheEntry struct {
 
 // perEPAllow maps EndpointIDs to ports + selectors + rules
-type perEPAllow map[uint64]portToSelectorAllow
+type perEPAllow map[uint64]portProtoToSelectorAllow
```
comment above should be updated to mention the protocol, as well as the one for `portProtoToSelectorAllow` below

pippolo84 commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -200,8 +200,12 @@ func asIPRule(r *regexp.Regexp, IPs map[string]struct{}) restore.IPRule {
+	if !exists {
+		return false
+	}
+	ipRules, exists := portMap[destPort]
```
Nit: you can also do

```
ipRules, exists := p.restored[endpointID][destProto][destPort]
if !exists {
  return false
}
```

directly, since reading from a nil map is safe.

pippolo84 commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -243,60 +247,66 @@ func (p *DNSProxy) GetRules(endpointID uint16) (restore.DNSRules, error) {
-						}).Warning("Too many IPs for a DNS rule, skipping the rest")
-						p.RUnlock()
-						break Loop
+	for proto, portMap := range portProtoToSelRegex {
```
Just a note that this method is becoming increasingly complex to read and understand.
We should consider a bit of refactoring here and in `RestoreRules` below too, but that's material for a follow-up PR.

pippolo84 commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -315,28 +325,30 @@ func (p *DNSProxy) RestoreRules(ep *endpoint.Endpoint) {
-				regex:  regex,
+	restoredRules := make(map[uint16]map[uint16][]restoredIPRule, len(ep.DNSRules))
+	for proto, portMap := range ep.DNSRules {
+		for port, dnsRule := range portMap {
```
Shouldn't we `make` the inner map here before ranging over `portMap`?

`restoredRules[proto] = make(map[uint16][]restoredIPRule, len(portMap))`

Otherwise the write at line 350 will panic.

pippolo84 commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -488,20 +513,27 @@ func (allow perEPAllow) setPortRulesForIDFromUnifiedFormat(cache regexCache, end
+		return nil, false
+	}
+	rules, exists = portMap[destPort]
 	return rules, exists
```
This can be simplified to:

```
	rules, exists = allow[endpointID][destProto][destPort]
	return
```

pippolo84 commented on pkg/fqdn/restore/restore.go:
```diff
@@ -48,10 +48,12 @@ func (r IPRules) Sort() IPRules {
-		if len(ipRules) > 0 {
-			ipRules = ipRules.Sort()
-			r[port] = ipRules
+	for proto, portMap := range r {
```
Not directly related to this PR, but if this should be used only for testing, we should consider adding a `*testing.T` parameter to its signature, in order to make it explicit that this is intended only for testing purposes (same for `IPRules.Sort`).

nathanjsweet replied:
Done.

pippolo84 reviewed (changes requested):
Thanks for the PR!

Note: the precheck error is due to a missing update of `MockFQDNProxy.UpdateAllowed` signature in `pkg/fqdn/proxy/proxy.go`.

nathanjsweet committed:
fqdn: Update DNS Restore to Index to PortProto

DNS Proxy needs to account for protocol when indexing
L7 DNS rules that it needs to adhere to, otherwise
L7 rules with differing port-protocols can override
each other (nondeterministically) and create overly
restrictive, and incorrect DNS rules. The problem with
accounting for protocol is that Endpoint restoration
logic uses DNS rules that index to port-only as JSON
saved to disk. Adding an additional protocol index to
a map structure changes the JSON structure and breaks
restoration logic between Cilium versions.

This change makes the map index backwards compatible,
since it changes the index from a uint16 to a uint32,
both of which marshal the same into a JSON structure.
The endpoint restoration logic will succeed between
versions, because the older version will be
automatically differentiated with a lack of a 1-bit
at bit position 24. Version 2 will save a 1 bit at the
24th bit going forward to differentiate when protocol
is indexed or not present.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

nathanjsweet committed:
fqdn: Fallback to Version 1 Port Lookups

In cases where a port-protocol is not present
in an restored port protocol, look up
up the Version 1 version of the PortoProto
in case a Version 1 PortProto was restored.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

christarazi commented on pkg/fqdn/restore/restore.go:
```diff
@@ -12,8 +12,54 @@ import (
+// no values at postions 16-31 as the original Version 1
+// was a uint16. Version 2 enforces a 1 value at the 24th
+// bit position, so it will alway be legible.
+type PortProto uint32
```
Clever!

christarazi reviewed (commented):
The backwards-compat restoration logic LGTM assuming manual validation. I only studied that given I was only focused on ensuring that upgrades are not broken in this area of the code (see https://github.com/cilium/cilium/pull/12964). I did not review the rest of the PR, so I will leave that to the existing assigned reviewers for that.

Also, it's worth mentioning that in future versions, we will not need the "migration" / version 1 code anymore and so we can remove it. Worth filing an issue so that we don't forget.

nathanjsweet committed:
testing: Update Restore Sort Method Signatrues

The Sort methods are updated to take an unused
testing.T structure to indicate to all callers
that they are only for testing purposes.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

pippolo84 reviewed (approved):
Thanks! :100:

nathanjsweet committed:
endpoint: Create a New Restore Field for DNS

DNSRulesV2 accounts for protocol and DNSRules does not.
DNSProxy needs to account for both, and endpoint needs
to be able to restore from a downgrade. DNSRulesV2 is used
by default now, but DNSRules is maintained in case of a
downgrade.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>
